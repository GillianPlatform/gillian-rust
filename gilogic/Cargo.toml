[package]
name = "gilogic"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
gilogic_proc = { path = "../gilogic_proc" }
gilogic_proc_dummy = { path = "../gilogic_proc_dummy" }