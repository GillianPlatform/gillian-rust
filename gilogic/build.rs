fn main() {
    println!("cargo::rustc-check-cfg=cfg(gillian)");
}
