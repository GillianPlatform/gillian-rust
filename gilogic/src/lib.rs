#![allow(internal_features)]
#![feature(rustc_attrs)]
#![feature(never_type)]
#![feature(register_tool)]
#![feature(ptr_internals)]
#![feature(allocator_api)]
#![feature(tuple_trait, unboxed_closures)]
#![feature(stmt_expr_attributes, proc_macro_hygiene)]
#![register_tool(gillian)]

mod tys;

pub use tys::RustAssertion;

pub mod macros {
    pub use gilogic_proc::{
        assertion, assertion_test, borrow, close_borrow, extract_lemma, lemma, open_borrow,
        predicate, show_safety, specification,
    };
    pub mod prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_pcy as with_freeze_lemma_for_mutref;
    }
    pub mod no_prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_no_pcy as with_freeze_lemma_for_mutref;
    }
}

mod seq;
pub use seq::Seq;

mod ownable;
pub use ownable::Ownable;
pub mod alloc;
pub mod iterated;
pub mod prophecies;

pub mod symex_ctrl {

    #[gillian::builtin]
    #[gillian::no_translate]
    #[rustc_diagnostic_item = "gillian::symex_ctrl::symex_next"]
    #[inline(always)]
    pub fn symex_next() {}
}

#[path = "stubs.rs"]
pub mod __stubs;
