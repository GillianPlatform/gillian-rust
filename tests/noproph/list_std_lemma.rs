//@check-pass
extern crate gilogic;

use gilogic::{macros::*, unfold, Ownable};
use std::marker::PhantomData;
use std::ptr::NonNull;

// 0 modification, although since I copied it, the allocator API was added
pub struct LinkedList<T> {
    head: Option<NonNull<Node<T>>>,
    tail: Option<NonNull<Node<T>>>,
    len: usize,
    marker: PhantomData<Box<Node<T>>>,
}

struct Node<T> {
    next: Option<NonNull<Node<T>>>,
    prev: Option<NonNull<Node<T>>>,
    element: T,
}

#[predicate]
fn dll_seg<T: Ownable>(
    head: In<Option<NonNull<Node<T>>>>,
    tail_next: In<Option<NonNull<Node<T>>>>,
    tail: In<Option<NonNull<Node<T>>>>,
    head_prev: In<Option<NonNull<Node<T>>>>,
    len: usize,
) {
    assertion!((head == tail_next) * (tail == head_prev) * (len == 0));
    assertion!(|hptr, head_next, head_prev, element|
        (head == Some(hptr)) *
        (hptr -> Node { next: head_next, prev: head_prev, element }) *
        element.own() *
        dll_seg(head_next, tail_next, tail, head, len - 1)
    )
}

#[predicate]
fn dll_seg_r<T: Ownable>(
    head: In<Option<NonNull<Node<T>>>>,
    tail_next: In<Option<NonNull<Node<T>>>>,
    tail: In<Option<NonNull<Node<T>>>>,
    head_prev: In<Option<NonNull<Node<T>>>>,
    len: usize,
) {
    assertion!((head == tail_next) * (tail == head_prev) * (len == 0));
    assertion!(|tptr, tail_next, tail_prev, element|
        (tail == Some(tptr)) *
        (tptr -> Node { next: tail_next, prev: tail_prev, element }) *
        element.own() *
        dll_seg_r(head, tail, tail_prev, head_prev, len - 1)
    )
}

#[lemma]
#[specification(
     forall len.
     requires {
         dll_seg(head, tail_next, tail, head_prev, len)
     }
     ensures {
         dll_seg_r(head, tail_next, tail, head_prev, len)
     }
 )]
fn dll_seg_l_to_r<T: Ownable>(
    head: Option<NonNull<Node<T>>>,
    tail_next: Option<NonNull<Node<T>>>,
    tail: Option<NonNull<Node<T>>>,
    head_prev: Option<NonNull<Node<T>>>,
) {
    // assert_bind!(len | dll_seg(head, tail_next, tail, head_prev, len));
    unfold!(dll_seg(head, tail_next, tail, head_prev, len));
}
