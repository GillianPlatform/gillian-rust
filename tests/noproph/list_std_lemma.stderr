[TIMESTAMP WARN  lib_rtg::metadata] could not read metadata for crate `"std"`: Os { code: 2, kind: NotFound, message: "No such file or directory" }
warning: unused variable: `next`
  --> ../tests/noproph/list_std_lemma.rs:72:5
   |
72 |     next: Option<NonNull<Node<T>>>,
   |     ^^^^ help: if this is intentional, prefix it with an underscore: `_next`
   |
   = note: `#[warn(unused_variables)]` on by default

warning: unused variable: `hptr`
  --> ../tests/noproph/list_std_lemma.rs:97:22
   |
97 |         assert_bind!(hptr, head_next, head_prev, element |
   |                      ^^^^ help: if this is intentional, prefix it with an underscore: `_hptr`

warning: unused variable: `head_prev`
  --> ../tests/noproph/list_std_lemma.rs:97:39
   |
97 |         assert_bind!(hptr, head_next, head_prev, element |
   |                                       ^^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_head_prev`

warning: unused variable: `element`
  --> ../tests/noproph/list_std_lemma.rs:97:50
   |
97 |         assert_bind!(hptr, head_next, head_prev, element |
   |                                                  ^^^^^^^ help: if this is intentional, prefix it with an underscore: `_element`

warning: fields `head`, `tail`, and `len` are never read
  --> ../tests/noproph/list_std_lemma.rs:10:5
   |
9  | pub struct LinkedList<T> {
   |            ---------- fields in this struct
10 |     head: Option<NonNull<Node<T>>>,
   |     ^^^^
11 |     tail: Option<NonNull<Node<T>>>,
   |     ^^^^
12 |     len: usize,
   |     ^^^
   |
   = note: `#[warn(dead_code)]` on by default

warning: fields `next`, `prev`, and `element` are never read
  --> ../tests/noproph/list_std_lemma.rs:17:5
   |
16 | struct Node<T> {
   |        ---- fields in this struct
17 |     next: Option<NonNull<Node<T>>>,
   |     ^^^^
18 |     prev: Option<NonNull<Node<T>>>,
   |     ^^^^
19 |     element: T,
   |     ^^^^^^^

warning: 6 warnings emitted

[TIMESTAMP DEBUG rust_to_gil] Correct!
